react/ExtensionContainer.tsx

```diff
@@ -1,26 +1,23 @@
 import PropTypes from 'prop-types'
 import React, {Component, Fragment} from 'react'
 
+import {RenderContext} from './components/RenderContext'
 import ExtensionPoint from './ExtensionPoint'
-import {getDirectChildren} from './utils/treePath'
+import {getDirectChildren, TreePathContext} from './utils/treePath'
 
 class ExtensionContainer extends Component {
-  public static contextTypes = {
-    extensions: PropTypes.object,
-    treePath: PropTypes.string,
-  }
-
-  public context!: RenderContext
-
   public render() {
-    const {extensions, treePath} = this.context
-    const children = getDirectChildren(extensions, treePath)
     return (
-      <Fragment>
-        {children.map(id =>
-          <ExtensionPoint {...this.props} key={id} id={id} />
-        )}
-      </Fragment>
+      <RenderContext.Consumer>
+        {runtime =>
+          <TreePathContext.Consumer>
+            {({treePath}) =>
+              getDirectChildren(runtime!.extensions, treePath)
```

Comment:
Actually, that will never happen. `React.createContext` requires an initial value but many of our context fields are only available when creating an instance of `RenderProvider`. So, it was a choice between a context that is always there (`{}`), but you need to check every field, and a context that can be undefined, but in practice it isn't, because `RenderProvider` is guaranteed be there and set its value.

I thought the latter was better:
```ts
export const RenderContext = React.createContext<RenderContext | undefined>(undefined)
```

But, I just figured out I can fool Typescript this way:
```ts
export const RenderContext = React.createContext<RenderContext>({} as any)
```

I'll do that and remove these `!.`